server/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -438,58 +435,57 @@ private void loadAllLookupsAndInitStateRef()
-                  LOG.info(
-                      "Starting lookup [%s]:[%s]",
-                      lookupBean.getName(),
-                      container
-                  );
-                  if (container.getLookupExtractorFactory().start()) {
-                    LOG.info(
-                        "Started lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return new AbstractMap.SimpleImmutableEntry<>(lookupBean.getName(), container);
-                  } else {
-                    LOG.error(
-                        "Failed to start lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return null;
-                  }
-                }
-            )
-        );
-      }
-      final ListenableFuture<List<Map.Entry>> futureList = Futures.allAsList(futures);
-      try {
-        futureList.get()
-                  .stream()
-                  .filter(Objects::nonNull)
-                  .forEach(builder::put);
+        completionService.submit(
+            () -> {
 
-        stateRef.set(new LookupUpdateState(builder.build(), ImmutableList.of(), ImmutableList.of()));
-      }
-      catch (InterruptedException | ExecutionException ex) {
-        futureList.cancel(true);
-        throw ex;
+              LookupExtractorFactoryContainer container = lookupBean.getContainer();
+              LOG.info(
+                  "Starting lookup [%s]:[%s]",
+                  lookupBean.getName(),
+                  container
+              );
+              if (container.getLookupExtractorFactory().start()) {
+                LOG.info(
+                    "Started lookup [%s]:[%s]",
+                    lookupBean.getName(),
+                    container
+                );
+                return new AbstractMap.SimpleImmutableEntry<>(lookupBean.getName(), container);
+              } else {
+                LOG.error(
+                    "Failed to start lookup [%s]:[%s]",
+                    lookupBean.getName(),
+                    container
+                );
+                return null;
+              }
+            }
+        );
       }
+      IntStream.range(0, lookupBeanList.size())
+               .forEach(i -> {
+                 try {
+                   final Future<AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer>> completedFuture = completionService
+                       .take();
+                   final AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer> lookupResult = completedFuture
+                       .get();
+                   if (lookupResult != null) {
+                     builder.put(lookupResult);
+                   }
+                 }
+                 catch (InterruptedException | ExecutionException e) {
```

Comment:
On InterruptedException, it should fall through to the `catch (Exception e) {` block below. And cancel all the remaining futures.

Reply:
I have cancelled the remaining futures on ExecutionException.